Auto-generate Conway CDDL using Cuddle

nc6 opened the pull request:
# Description

This PR resolves #4125 by adding a new conway executable utility, `huddle-cddl`, which will auto-generate the Conway CDDL (as a single file) from the Huddle specification. Consequently, the existing specifications have been removed.

The existing CDDL tests have been remapped to be based upon the generated file and confirmed to work correctly.

In addition, it resolves the remaining comments from #4437, namely:
 
- The issue with `/` and `//` has been resolved. There is now only the `/` operator in Huddle, and its usage is carefully documented.
- `header` has been reformatted
- Deprecation warnings have been added to `stake_registration`, `stake_deregistration`, and `redeemers`.
- Duplicate comments have been removed
- The "n" argument for `script_n_of_k` has now been properly sized as an `int64`.


# Checklist

- [x ] Commit sequence broadly makes sense and commits have useful messages
- [x ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

nc6 committed:
Add utility to generate cardano CDDL files.

This addresses issue #4461

We add two things:
- A haskell tool to generate the CDDL files
- A script which verifies if any Cuddle specs have
  changed, and invokes the tool if so.

Both of these would need to be updated in case
additional Cuddle specs were added.

The tool can be invoked with `cabal run cardano-cddl`.

nc6 committed:
Generate CDDL files from Huddle.

This additionally bumps the Huddle version to
include a change to how choices are processed.

nc6 committed:
Update flake.lock

nc6 committed:
Bump chap index-state

Needed to pull in the latest release of cuddle

lehins commented on eras/conway/impl/cddl-files/conway-merged.cddl:
Why is it called `conway-merged.cddl`? Could you please rename it to `conway.cddl`.

nc6 replied:
Done

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -82,7 +87,7 @@ protocol_version :: Rule
 
 -- TODO Replace with the following once
--- https://github.com/input-output-hk/cuddle/issues/29 is addressed in cuddle.
+-- https:/github.com/input-output-hk/cuddle/issues/29 is addressed in cuddle.
```
Looks like search and replace gone wrong :smile: 
```suggestion
-- https://github.com/input-output-hk/cuddle/issues/29 is addressed in cuddle.
```

nc6 replied:
Whoops

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -770,23 +772,23 @@ policy_id = "policy_id" =:= scripthash
 asset_name = "asset_name" =:= VBytes `sized` (0 :: Word64, 32 :: Word64)
 
--- Once https://github.com/input-output-hk/cuddle/issues/29 is in place, replace
+-- Once https:/github.com/input-output-hk/cuddle/issues/29 is in place, replace
```
Ditto
```suggestion
-- Once https://github.com/input-output-hk/cuddle/issues/29 is in place, replace
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -770,23 +772,23 @@ policy_id = "policy_id" =:= scripthash
 minInt64 = -9223372036854775808
 
--- Once https://github.com/input-output-hk/cuddle/issues/29 is in place, replace
+-- Once https:/github.com/input-output-hk/cuddle/issues/29 is in place, replace
```
```suggestion
-- Once https://github.com/input-output-hk/cuddle/issues/29 is in place, replace
```

lehins commented on eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs:
```diff
@@ -770,23 +772,23 @@ policy_id = "policy_id" =:= scripthash
 maxInt64 = 9223372036854775807
 
--- Once https://github.com/input-output-hk/cuddle/issues/29 is in place, replace
+-- Once https:/github.com/input-output-hk/cuddle/issues/29 is in place, replace
```
```suggestion
-- Once https://github.com/input-output-hk/cuddle/issues/29 is in place, replace
```

lehins commented on libs/cardano-cddl/cardano-cddl.cabal:
```diff
@@ -0,0 +1,34 @@
+cabal-version:      3.4
+name:               cardano-cddl
```
I'd rather avoid adding new packages. Especially tiny packages like this one.

I relive it would be better for `cardano-ledger-conway` to just get an executable section `gen-ccdl` instead, especially since it suppose to generate a file that is part of `cardano-ledger-conway` package itself

nc6 replied:
The idea behind a new package is that there would be a single tool to generate the CDDL for all eras, rather than needing one additional binary per era. It's just that right now, only Conway has been converted. Thoughts?

lehins replied:
I understand the idea. I am saying that this idea is wrong and doesn't fit with how we structure everything in `cardano-ledger` repo. We do want each era package to be responsible for everything that is going on in that era, especially things like cddl generation. 

So, yes, please create a new executable in `cardano-ledger-conway` instead of a separate package. Later all other eras will get the same kind of executable.

lehins commented on eras/conway/impl/cddl-files/conway-merged.cddl:
```diff
@@ -0,0 +1,506 @@
+
+auth_committee_hot_cert = (14
+                           , committee_cold_credential
+                           , committee_hot_credential)
```
Can we improve pretty printer a bit to format it a bit nicer? Feel free to just create tickets on cuddle for later for formatting improvements
```suggestion
auth_committee_hot_cert = ( 14
                          , committee_cold_credential
                          , committee_hot_credential
                          )
```

nc6 replied:
Yep, the pretty printer right now is pretty basic, can definitely be improved!

nc6 replied:
https://github.com/input-output-hk/cuddle/issues/34

lehins commented on eras/conway/impl/cddl-files/conway-merged.cddl:
```diff
@@ -0,0 +1,506 @@
+                            , ? 1 : [* native_script]
+                            , ? 2 : [* plutus_v1_script]
+                            , ? 3 : [* plutus_v2_script]
+                            , ? 4 : [* plutus_v3_script]})
```
This is another formatting improvement suggestion:
```suggestion
                  / #6.259( { ? 0 : metadata
                            , ? 1 : [* native_script]
                            , ? 2 : [* plutus_v1_script]
                            , ? 3 : [* plutus_v2_script]
                            , ? 4 : [* plutus_v3_script]
                            }
                          )
```

lehins commented on eras/conway/impl/cddl-files/conway-merged.cddl:
```diff
@@ -0,0 +1,506 @@
+         , transaction_bodies : [* transaction_body]
+         , transaction_witness_sets : [* transaction_witness_set]
+         , auxiliary_data_set : {* transaction_index => auxiliary_data}
+         , invalid_transactions : [* transaction_index]]
```
Ditto:
```suggestion
block = [ header
        , transaction_bodies : [* transaction_body]
        , transaction_witness_sets : [* transaction_witness_set]
        , auxiliary_data_set : {* transaction_index => auxiliary_data}
        , invalid_transactions : [* transaction_index]
        ]
```

lehins commented on eras/conway/impl/cddl-files/conway-merged.cddl:
```diff
@@ -0,0 +1,506 @@
+
+vrf_keyhash = $hash32
+
+withdrawals = {+ reward_account => coin}
```
Generate a new line at the end of the file

lehins commented on scripts/gen-cddl.sh:
```diff
@@ -0,0 +1,9 @@
+changed=$(git diff --name-only origin/master -- 'eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/CDDL.hs')
+
+if [[ -n "$changed" ]]; then
+  cabal run cardano-cddl
```
`gen` prefix could be a bit ambiguous and overused, so maybe alternative names for the executable could be `cuddle-cddl` or `huddle-cddl`.
```suggestion
  cabal run cardano-ledger-conway:gen-cddl
```

lehins commented on libs/cardano-cddl/app/Main.hs:
```diff
@@ -0,0 +1,34 @@
+-- | Map from spec to the CDDL file location
+cuddleSpecs :: [(Huddle, String)]
+cuddleSpecs =
+  [ (Conway.conway, "./eras/conway/impl/cddl-files/conway-merged.cddl")
```
One more benefit of sticking this executable of `cardano-ledger-conway` is that you can use [`getDataFileName`](https://github.com/IntersectMBO/cardano-ledger/blob/88d788500ceb2b4bc5215afb6ebe6fd73972baf1/eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Binary/Cddl.hs#L11) in order to get the accurate path.

lehins commented on libs/cardano-cddl/app/Main.hs:
```diff
@@ -0,0 +1,34 @@
+  let cddl = toCDDL hddl
+   in withFile path WriteMode $ \h -> do
+        hPutStrLn h preface
+        hPutDoc h (pretty $ sortCDDL cddl)
```
Why do we sort CDDL? I would really like the order of definition from the original Huddle file to be preserved.
```suggestion
        hPutDoc h (pretty cddl)
```

Messing up the order prevents us form logically grouping related types together

nc6 replied:
Do you know a way to do this? Currently, the only effect of this sorting is to have a single alphabetical list, rather than a list that has first the "rules" and then the "groups" (in CDDL these aren't distinguished, but it's helpful to do so in Huddle).

I guess perhaps using template Haskell might allow finding the actual order in the Huddle file?

lehins replied:
I see ... So if we don't sort the huddle we'll first get this order first

```haskell
[block, transaction]
      <> [kes_signature, language, potential_languages, signkeyKES]
```

followed or interleaved by all other reusable definitions.

Some sort of template haskell approach would probably work. However, it looks tricky, so unless you can easily figure out how to do it, we probably shouldn't worry about it at this point. In other words, if it looks like it'll be too much work then just create an issue on the cuddle repo about it and we can look into it at some point later time permitting

nc6 replied:
https://github.com/input-output-hk/cuddle/issues/35

lehins commented on libs/cardano-cddl/app/Main.hs:
```diff
@@ -0,0 +1,34 @@
+  [ (Conway.conway, "./eras/conway/impl/cddl-files/conway-merged.cddl")
+  ]
+
+writeSpec :: Huddle -> String -> IO ()
```
Any reusable helper functions like this can go into `cardano-ledger-binary:testlib`

lehins commented on libs/cardano-cddl/app/Main.hs:
```diff
@@ -0,0 +1,34 @@
+{-# LANGUAGE ImportQualifiedPost #-}
```
Let's not use this extension for consistency sake. If we are to switch to this style, then we would do this for the whole codebase, not just individual modules.
```suggestion
```

lehins reviewed (changes requested):
This looks great. One major thing I don't like is addition of a new package: it is totally unnecessary.

nc6 committed:
Address review comments

- Rename conway-merged.cddl to conway.cddl
- Add a newline at the end of generated CDDL files
- Remove the use of ImportQualifiedPost
- Fix the search/replace failure with https://
- Format cabal files using cabal-format

nc6 committed:
Move CDDL generation into cardano-ledger-conway

Each era will now have its own executable for generating CDDL files.

A utility function (writeSpec) has been moved into
cardano-ledger-binary:testlib for this purpose.

lehins commented on libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/Cuddle.hs:
```diff
@@ -200,3 +205,18 @@ cddlFailure encoding err =
+--------------------------------------------------------------------------------
+
+-- | Write a Huddle specification to a file at the given path
+writeSpec :: Cuddle.Huddle -> FilePath -> IO ()
```
Addition of a new function to `cardano-ledger-binary` requires a minor version bump and a changelog entry. So, could you please change the version to `1.3.4.0` and add a lower bound on it in `cardano-ledger-conway`, since that function is being used there.

lehins commented on eras/conway/impl/cardano-ledger-conway.cabal:
```diff
@@ -166,6 +164,21 @@ library testlib
+    build-depends:
+        base,
+        testlib,
+        cardano-ledger-binary:testlib
```
This is to ensure we don't try to build it with older `cardano-ledger-binary` that doesn't have `writeSpec`
```suggestion
        cardano-ledger-binary:testlib >= 1.3.4.0
```

lehins reviewed (approved):
This looks beautiful! Thank you!

nc6 committed:
Update version of cardano-ledger-binary

- Bump c-l-b to 1.3.4.0
- Add changelog entries
- Add lower bound in cardano-ledger-binary to reflect the use of
writeSpec
